src/state/_common/reducers.js

```diff
@@ -17,112 +29,134 @@ export const DEFAULT_INITIAL_STATE = {
+	 * @return {Object} updated state
+	 */
 	add: (state, action) => {
-		let newData = {...state.byKey};
-		if (action.data && action.data.length) {
+		if (action?.data?.length) {
+			let newData = {...state.byKey};
 			action.data.forEach(model => {
 				newData[model.key] = {...newData[model.key], ...model};
 				delete newData[model.key].outdated;
 				delete newData[model.key].unreceived;
 			});
-		}
-		return {...state, byKey: newData};
-	},
 
-	addBatch: (state, action) => {
-		let newData = {...state.byKey};
-		if (action.data && action.data.length) {
-			action.data.forEach(model => {
-				newData[model[action.key]] = {...newData[model[action.key]], ...model};
-				delete newData[model[action.key]].outdated;
-				delete newData[model[action.key]].unreceived;
-			});
+			return {...state, byKey: newData};
+		} else {
+			return state;
 		}
-		return {...state, byKey: newData};
 	},
 
+	/**
+	 * Add just keys and mark as unreceived for unreceived models
+	 * @param state {Object}
+	 * @param action {Object}
+	 * @param action.keys {Array} A list of unreceived keys
+	 * @return {Object} updated state
+	 */
 	addUnreceivedKeys: (state, action) => {
-		let newData = {...state.byKey};
-		if (action.keys && action.keys.length) {
+		if (action?.keys?.length) {
+			let newData = {...state.byKey};
 			action.keys.forEach(key => {
 				newData[key] = {key, unreceived: true};
 			});
+			return {...state, byKey: newData};
+		} else {
+			return state;
 		}
-		return {...state, byKey: newData};
 	},
 
+	/**
+	 * Add new or updated existing index
+	 * @param state {Object}
+	 * @param action {Object}
+	 * @param action.data {Array} A collection of models to add
+	 * @param action.filter {Object}
+	 * @param action.order {Array}
+	 * @param action.start {number}
+	 * @param action.length {number}
+	 * @param action.count {number}
+	 * @param action.changedOn {string}
+	 * @return {Object} updated state
+	 */
 	addIndex: (state, action) => {
-		let indexes = [];
-		let selectedIndex = {};
-
-		if (state.indexes) {
-			state.indexes.forEach(index => {
-				if (
-					_.isEqual(index.filter, action.filter) &&
-					_.isEqual(index.order, action.order)
-				) {
-					selectedIndex = index;
-				} else {
-					indexes.push(index);
+		if (action?.data) {
+			const updatedIndexes = state.indexes ? [...state.indexes] : [];
+			const indexOfIndex = updatedIndexes.length
+				? _findIndex(state.indexes, index => {
+						return (
+							_isEqual(index.filter, action.filter) &&
+							_isEqual(index.order, action.order)
+						);
+				  })
+				: -1;
+
+			// update existing index
+			if (indexOfIndex > -1) {
+				const updatedIndex = {...updatedIndexes[indexOfIndex]};
+
+				// register models to index
+				const updatedIndexIndex = commonHelpers.registerModelsToIndex(
+					{...updatedIndex.index},
+					action.data,
+					action.start
+				);
+
+				// Remove loading indicator if data does not come
```

Comment:
It happens when we ask for {offset:0, limit:10}, but from BE come only five items, because there is no more. We do not know, how many results if on BE at the moment when we ask, so we create index with loading indicator for all 10 items.

I can say that delete records is enough, everything works.

Reply:
Oh, right. Probably clarify the comment to specify it's not as much "data does not come" as "data on this position doesn't exist" or something similar.

Reply:
ef95685c467e24550a1c68c501d9503d4310f45e